crates/sui-graphql-rpc/src/types/string_input.rs

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Opt-in to an implementation of `ScalarType` for a `$Type` that implements `FromStr`, solely for
+/// use as an input (not an output). The type masquarades as a `String` in the GraphQL schema, to
+/// avoid adding a new scalar.
+macro_rules! impl_string_input {
+    ($Type:ident) => {
+        #[Scalar(name = "String", visible = false)]
+        impl ScalarType for $Type {
+            fn parse(value: Value) -> InputValueResult<Self> {
+                if let Value::String(s) = value {
+                    Ok(Self::from_str(&s)?)
+                } else {
+                    Err(InputValueError::expected_type(value))
+                }
+            }
+
+            fn to_value(&self) -> Value {
+                unimplemented!("String inputs should not be output");
```

Comment:
this would panic right? should we instead have this be an internal error? on the off-chance that someone does use this and it isn't caught for whatever reason

Reply:
It does panic -- we can't return an internal error unfortunately, because this is implementing an interface and `to_value`'s signature does not allow it to fail, so the trade-off is between trying to implement a version of `to_value` that works in all cases, or panicking.

It's possible to write such an implementation, e.g. by assuming `$Type` implements `fmt::Display` or `ToString`, but the reason I didn't like that was:

- It means that we need to come up with a way to output every scalar we want to do validation/parsing on, on input, which is a spurious requirement.
- We don't easily detect when types we intended only for input have leaked to output -- we may have chosen some dummy implementation of `Display` to appease the type checker, but we never hear about it when accidentally gets used.  This code panics reliably if you ever try and use this type in an output and write a test or do a local run that outputs the offending type, so for this panic to cause an issue in prod would require us to use the type incorrectly, and then not even bother to look at the output before landing it.

Ideally `async-graphql` would have a trait just for types we want to deserialize (you would think this was `InputType`, but from what I can tell, it also requires an infallible `to_value` implementation), but they don't so this is the next best way to work around that.